Add profiling tips and tricks to documentation (#3739)
input-output-hk/ouroboros-network: ouroboros-consensus/docs/Contributing.md

[consensus-style]: StyleGuide.md

[iohk-onboarding]: IOHKOnboarding.md

## Profiling tips & tricks

Here there are some findings or tricks we have found useful at different times
regarding profiling the Ouroboros Consensus layer.

### Time reported in GHC `-p` profiling is off

When a Haskell program is run, the reported time on the `-p` report is the result of

    total_ticks * tick_interval / num_capabilities
    
The Cardano-node is run by default with two capabilities, this means that the
total time will be half of what one would expect.

### Time reported is lower than half of what one would expect

> This can be due to the fact that there are `safe` and `unsafe` calls in

Reviewer: ```suggestion
This can be because there are `safe` and `unsafe` calls in
```

Reviewer: Maybe we should link to what safe and unsafe are.

running. 

Imagine the following code in Haskell:

```haskell
foreign import ccall unsafe "foo" foo :: IO ()
foreign import ccall safe   "bar" foo :: IO ()

baz = do
  foo
  bar
```

This will be compiled to Core (`-ddump-simpl`) as:

```
...
 case {__pkg_ccall main State# RealWorld -> (# State# RealWorld #)}_d20O
                        ds_X21f
                 of
...
case {__pkg_ccall_GC main State# RealWorld -> (# State# RealWorld #)}_d20S
             ipv4_X284
      of
...
```

And to C-- (`-ddump-cmm`) as:

```
...
           call "ccall" arg hints:  []  result hints:  [] foo();
...
           (_u3jD::I64) = call "ccall" arg hints:  [PtrHint,]  result hints:  [PtrHint] suspendThread(BaseReg, 0);
           call "ccall" arg hints:  []  result hints:  [] bar();
           (_u3jE::I64) = call "ccall" arg hints:  [PtrHint]  result hints:  [PtrHint] resumeThread(_u3jD::I64);
...
```

This shows how a `safe` call indeed releases the capability. Imagining a super
simplified scenario with initially one OS thread, one Haskell task/thread and
only one capability that performs the `baz` call:

```
= has the capability

> - not

Reviewer: Has not the capability?

>            v call to `foo`      v call to `bar`

Reviewer: Does the plot depend on the font used? The "arrows" seem off.

...===============================-------------------========= ...
                              ^ return             ^ return
```

So in the above, the time during the `safe` call, the Haskell capability is
`IDLE` thus not counting for the time spent. What happens if we have more
Haskell tasks waiting to be executed (note this will require multiple OS
threads)?

```
= has the capability
- not

           v call to `foo`      v call to `bar`
...===============================----------------------========= ...
                              ^ return             ^ return
...-----------------------------========================-------- ...
```

Note now that the foreign call might finish before the capability is available

> again, and in this case it has to wait.

Reviewer: What would the profile report if the capability has to wait? Will it report the waiting time?

--------------------------------------------------------------------------------

Now, the reported time on the Haskell time profiling will be off due to these
`safe` foreign calls not being accounted for. Apart from that, `unsafe` calls
sometimes lose ticks when running with the default tick interval. Therefore it

> is advised to run with `-V0.01` when doing measurements on time.

Reviewer 2: FTR: I opened https://gitlab.haskell.org/ghc/ghc/-/merge_requests/8186 to correct the incorrect default value mentioned [here](https://downloads.haskell.org/ghc/8.10.7/docs/html/users_guide/profiling.html#rts-flag--V%20%E2%9F%A8secs%E2%9F%A9) as discovered by Nick.

### Time profiling and eventlog

> One of the few approaches that seemed useful was using

Reviewer: Useful for?

[hs-speedscope](https://github.com/mpickering/hs-speedscope) to graph all the
execution time registered by Haskell, and then combining it with
[ghc-events-analyze](https://github.com/well-typed/ghc-events-analyze) to see
time elapsed between different events that we care of.

In order to emit time profiling on the eventlog, the executable has to be linked
with `-eventlog` to pull the right RTS and then run with `-l`. If the eventlog
produced is too big, one can run with `-l-au` so that (`-a`) all events are
omitted, (`u`) except the user events. Also one can set the output name for the

> eventlog file with `-ol<name>`.

Reviewer: I wonder if we could include some examples of how to invoke this.
